src/api/VoteMonitor.Api.Form/Mappers/HierarchicalMapper.cs

```diff
@@ -0,0 +1,71 @@
+﻿using AutoMapper;
+using System.Collections.Generic;
+using System.Linq;
+using VoteMonitor.Entities;
+
+namespace VoteMonitor.Api.Form.Mappers
+{
+    /// <summary>
+    /// Represents a helper class used to map an hierarchic entity from its DTO by either create or update.
+    /// </summary>
+    public class HierarchicalMapper<TEntity, TDto, TChildEntity, TChildDto> : IEntityMapper<TEntity, TDto>
+        where TEntity : IHierarchicalEntity<TChildEntity>
+        where TDto : IHierarchicalEntity<TChildDto>
+        where TChildEntity : IIdentifiableEntity
+        where TChildDto : IIdentifiableEntity
+    {
+        private readonly IUpdateOrCreateEntityMapper<TEntity, TDto> _updateOrCreateEntityMapper;
+        private readonly IEntityMapper<TChildEntity, TChildDto> _updateOrCreateChildEntityMapper;
+
+        public HierarchicalMapper(IUpdateOrCreateEntityMapper<TEntity, TDto> updateOrCreateEntityMapper, IEntityMapper<TChildEntity, TChildDto> updateOrCreateChildEntityMapper)
+        {
+            _updateOrCreateEntityMapper = updateOrCreateEntityMapper;
+            _updateOrCreateChildEntityMapper = updateOrCreateChildEntityMapper;
+        }
+
+        public void Map(ref TEntity entity, TDto dto)
+        {
+            // We map the entity from dto using the update or create entity mapper.
+            _updateOrCreateEntityMapper.Map(ref entity, dto);
+
+            // In case the children property is null, we'll initialize it with an empty list.
+            entity.Children ??= new List<TChildEntity>();
+
+            // We get the left join from the dto's childtren and the entity's children
+            var childData = from childDto in dto.Children ?? Enumerable.Empty<TChildDto>()
+                            join child in entity.Children
+                            on childDto.Id equals child.Id
+                            into childDataJoined
+                            from child in childDataJoined.DefaultIfEmpty()
+                            select new { Child = child, ChildDto = childDto };
+
+            var childrenToAdd = new List<TChildEntity>();
+
+            foreach (var (childDataItem, childIndex) in childData.Select((childDataItem, childIndex) => (childDataItem, childIndex + 1)))
+            {
+                var child = childDataItem.Child;
+
+                // We update the child property using the template method.
+                _updateOrCreateChildEntityMapper.Map(ref child, childDataItem.ChildDto);
+
+                // If the entity is ordered, we set the childIndex on it.
+                if (child is IOrderedEntity orderedChildEntity)
+                {
+                    orderedChildEntity.OrderNumber = childIndex;
```

Comment:
order number should be retrieved from model posted to that endpoint

Reply:
Hi @idormenco,
In the previous implementation both for add and update, the OrderNumber was generated using the index of the entity. It makes sense to be updated from DTO but I'm wondering to not impact any business logic(e.g. some client is not sending the order number, etc).
Maybe @aniri can confirm that the OrderNumber should be retrieved from the DTO. In the meantime, I'll make these changes as it makes sense to me too (I'll also remove IOrderedEntity as it will be no longer used).

Reply:
Yep, the order number should be retrieved from the DTO :) 

Thank you!